directives/animate/animate.js

```diff
@@ -0,0 +1,250 @@
+import { directive, PropertyPart } from 'lit-html';
+import { getComposedActiveElement, getNextFocusable } from '../../helpers/focus.js';
+import { getComposedParent, isComposedAncestor } from '../../helpers/dom.js';
+
+const stateMap = new WeakMap();
+const reduceMotion = matchMedia('(prefers-reduced-motion: reduce)').matches;
+const showTransitionDuration = 300;
+const hideTransitionDuration = 200;
+const moveYValue = 20;
+
+class AnimationState {
+
+	constructor(propertyPart) {
+
+		if (!(propertyPart instanceof PropertyPart) || propertyPart.committer.name !== 'animate') {
+			throw new Error('animation directives must be used with "animation" property');
+		}
+
+		this.id = 0;
+		this.clone = null;
+		this.elem = propertyPart.committer.element;
+		this.state = 'unknown';
+
+	}
+
+	async animate(info) {
+
+		if (this.clone === null) {
+			this.clone = document.createElement('div');
+		}
+		this.elem.parentNode.insertBefore(this.clone, this.elem);
+
+		const id = ++this.id;
+
+		let outerResolve;
+		const onTransitionEnd = (event) => {
+			if (event.target === this.clone) {
```

Comment:
When would the target not be the clone?  If it will always be the clone, perhaps we can use the `{once: true}` option which will cleanup the event handler automatically.  Note however, `transitionend` will be dispatched once for each property being animated, but it looks like we only care to run this handler once.

I'm also wondering if we need to consider the same cleanup for the `transitioncancel` event case.  🤔   For that case, I think it would leave the clone in the DOM, and the outer promise would never resolve.  I kinda want to play with this a little bit, since I'm not certain of the behavior of the `show` is run, but then `hide` is run before `show` completes, and vice versa.

Reply:
The target-clone check was borrowed from Andy's code, not sure if it's necessary.

The code here that's checking the `id` value is meant to ensure that only the most recent transition's code will run the cleanup. So you can hide/show like crazy (I tested this by slowing the animations down) and the cleanup doesn't happen until the last one finishes.

There may be better ways to handle this with the `transitioncancel` and maybe I should be checking for the different properties. I can look into that a bit more.

Reply:
Originally, the target-clone check was in the mixin code because the clone wrapped around the content. `transitionend` events fired from the content would bubble up into the clone's listener, and cause the clone to do things when the content's transition ended, rather than when the clone's transition ended. This was especially apparent because of the slight delay between clone and content transitions.

Reply:
I think sometimes `transitionend` may not run is `transitioncancel` runs instead... but I'm not 100% clear on when `transitioncancel` runs.  I know it's not an event that we wire up very often though... I think we've sorta been ignoring it.  😬

Reply:
Here's what I've learned about `transitioncancel`:

- `transitioncancel` is only possible if the animation is delayed or has a duration (ours has a duration)
- only one of `transitioncancel` OR `transitionend` will run
- `transitioncancel` only runs if the animation is stopped before it completes. The scenarios under which this could happen are typically where the transition itself is only being applied under certain scenarios -- like while the user is hovering or while a particular element has a CSS class. As soon as those stop being true, if the tranistion hasn't finished, it'll be cancelled.

Since we're the ones controlling when the `transition` attribute is present, and we only remove it once everything is finished, I think we're OK to ignore it.